pdf/model/textencoding/simple.go

```diff
@@ -0,0 +1,1735 @@
+// The bool return flag is true if there was a match, and false otherwise.
+func (se SimpleEncoder) RuneToCharcode(val rune) (uint16, bool) {
+	return doRuneToCharcode(se, val)
+}
+
+// CharcodeToRune returns the rune corresponding to character code `code`.
+// The bool return flag is true if there was a match, and false otherwise.
+func (se SimpleEncoder) CharcodeToRune(code uint16) (rune, bool) {
+	r, ok := se.codeToRune[code]
+	if !ok {
+		common.Log.Debug("Charcode -> Rune error: charcode not found: 0x%04x", code)
+	}
+	return r, ok
+}
+
+// RuneToGlyph returns the glyph corresponding to rune `r`.
+// The bool return flag is true if there was a match, and false otherwise.
+func (se SimpleEncoder) RuneToGlyph(r rune) (string, bool) {
+	return runeToGlyph(r, glyphlistRuneToGlyphMap)
+}
+
+// GlyphToRune returns the rune corresponding to glyph `glyph`.
+// The bool return flag is true if there was a match, and false otherwise.
+func (se SimpleEncoder) GlyphToRune(glyph string) (rune, bool) {
+	return glyphToRune(glyph, glyphlistGlyphToRuneMap)
+}
+
+// ToPdfObject returns `se` as a PdfObject
+func (se SimpleEncoder) ToPdfObject() core.PdfObject {
+	if se.differences == nil {
+		return core.MakeName(se.baseName)
+	}
+	dict := core.MakeDict()
+	dict.Set("Type", core.MakeName("Encoding"))
+	dict.Set("BaseEncoding", core.MakeName(se.baseName))
+	dict.Set("Differences", core.MakeArray(ToFontDifferences(se.differences)...))
+	return core.MakeIndirectObject(dict)
+}
+
+// makeEncoder returns a SimpleEncoder based on `codeToRune`.
+func (se *SimpleEncoder) makeEncoder() {
+	codeToRune := map[uint16]rune{}
+	for code, r := range se.baseEncoding {
+		codeToRune[code] = r
+	}
+	if se.differences != nil {
+		for code, glyph := range se.differences {
+			r, ok := GlyphToRune(glyph)
+			if !ok {
+				common.Log.Debug("ERROR: No match for glyph=%q differences=%+v", glyph, se.differences)
+			}
+			codeToRune[uint16(code)] = r
+		}
+	}
+
+	codeToGlyph := map[uint16]string{}
+	glyphToCode := map[string]uint16{}
+	for code, r := range codeToRune {
+		glyph := glyphlistRuneToGlyphMap[r] // !@#$ Build out this map
+		codeToGlyph[code] = glyph
+		glyphToCode[glyph] = code
+		if glyph == "" {
+			common.Log.Debug("ERROR: Empty glyph code=0x%04x r=%+q=%#q", code, r, r)
+		}
+	}
+	se.codeToGlyph = codeToGlyph
+	se.glyphToCode = glyphToCode
+	se.codeToRune = codeToRune // XXX: !@#$ Make this a string
+}
+
+func FromFontDifferences(diffList []core.PdfObject) (map[byte]string, error) {
+	differences := map[byte]string{}
+	var n byte
+	for _, obj := range diffList {
+		switch v := obj.(type) {
+		case *core.PdfObjectInteger:
+			n = byte(*v)
+		case *core.PdfObjectName:
+			s := string(*v)
+			differences[n] = s
+			n++
+		default:
+			common.Log.Debug("ERROR: Bad type. obj=%s", obj)
+			return nil, core.ErrTypeError
+		}
+	}
+	return differences, nil
+}
+
+func ToFontDifferences(differences map[byte]string) []core.PdfObject {
```

Comment:
godoc comment
// ToFontDifferences converts a mapping of charcodes to glyphs difference list to a list of PdfObjects for use in PDF output.

Reply:
Done